extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -0,0 +1,571 @@
+                        job,
+                        KubernetesTaskRunnerConfig.toMilliseconds(k8sConfig.k8sjobLaunchTimeout),
+                        TimeUnit.MILLISECONDS
+                    );
+                    log.info("Job %s launched in k8s", k8sTaskId);
+                    completedPhase = monitorJob(peonPod, task, k8sTaskId);
+                  } else {
+                    Job job = existingJob.get();
+                    if (job.getStatus().getActive() == null) {
+                      if (job.getStatus().getSucceeded() != null) {
+                        completedPhase = new JobResponse(job, PeonPhase.SUCCEEDED);
+                      } else {
+                        completedPhase = new JobResponse(job, PeonPhase.FAILED);
+                      }
+                    } else {
+                      // the job is active lets monitor it
+                      completedPhase = monitorJob(task, k8sTaskId);
+                    }
+                  }
+                  TaskStatus status;
+                  if (PeonPhase.SUCCEEDED.equals(completedPhase.getPhase())) {
+                    status = TaskStatus.success(task.getId(), completedPhase.getLocation());
+                  } else {
+                    status = TaskStatus.failure(
+                        task.getId(),
+                        "Task failed %s: " + k8sTaskId
+                    );
+                  }
+                  if (completedPhase.getJobDuration().isPresent()) {
+                    status = status.withDuration(completedPhase.getJobDuration().get());
+                  }
+                  updateStatus(task, status);
+                  return status;
+                }
+                catch (Exception e) {
+                  log.error(e, "Error with task: %s", k8sTaskId);
+                  throw e;
+                }
+                finally {
+                  // publish task logs
+                  Path log = Files.createTempFile(task.getId(), "log");
+                  try {
+                    FileUtils.write(
+                        log.toFile(),
+                        client.getJobLogs(new K8sTaskId(task.getId())),
+                        StandardCharsets.UTF_8
+                    );
+                    taskLogPusher.pushTaskLog(task.getId(), log.toFile());
+                  }
+                  finally {
+                    Files.deleteIfExists(log);
+                  }
+                  client.cleanUpJob(new K8sTaskId(task.getId()));
+                  synchronized (tasks) {
+                    tasks.remove(task.getId());
+                  }
+                }
+              })
+          ));
+      return tasks.get(task.getId()).getResult();
+    }
+  }
+
+  JobResponse monitorJob(Task task, K8sTaskId k8sTaskId)
+  {
+    return monitorJob(client.getMainJobPod(k8sTaskId), task, k8sTaskId);
+  }
+
+  JobResponse monitorJob(Pod peonPod, Task task, K8sTaskId k8sTaskId)
```

Comment:
Might it be a bit easier to follow if, when moving the "controller" to a separate class, these methods were members of that new class? These seem more related to the controller (which would know the pod, task and K8s task id) than to the runner (which, it seems, is the thing that launches the controller.)